### PR TITLE
 chore: automated unit testing with jest

### PR DESCRIPTION
**Pull request by dlouzan**
- add initial unit testing based on jest as verdaccio
- minor refactor of authcache

### CODE REVIEW

File: test/unit/authcache.spec.js [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    authCache.storeUser(user, pass, data);
+    const returnedData: UserData = authCache.findUser(user, pass);
+
+    expect(returnedData).toEqual(data);
+  });
+
+  test('should store and find some user data when ttl is unlimited', () => {
+    const UNLIMITED_TTL: number = 0;
+    const authCache: AuthCache = new AuthCache(logger, UNLIMITED_TTL);
+    const user: string = 'fooUser';
```
Comment (ercanucan) [inline test/unit/authcache.spec.js]: minor: Would it make sense to have test user, pass, groups etc. as a constant for these tests? then you would not need to duplicate code.

File: test/unit/gitlab.spec.js [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+    const cb: Callback = (err, data) => {
+      expect(err).toBeTruthy();
+      expect(data).toBeFalsy();
+      done();
+    };
+
+    verdaccioGitlab.authenticate(user, pass, cb);
+  });
+
+  test('should fail authentication with not existing user', done => {
```
Comment (ercanucan) [inline test/unit/gitlab.spec.js]: minor: `non-existing user`

File: test/unit/gitlab.spec.js [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+      done();
+    };
+
+    verdaccioGitlab.allow_publish(user, _package, cb);
+  });
+
+  test('should deny publish of package based on user', done => {
+    const verdaccioGitlab: VerdaccioGitlab = new VerdaccioGitlab(defaultConfig, options);
+    const user: RemoteUser = {
+      real_groups: ['myGroup', 'myUser'],
```
Comment (ercanucan) [inline test/unit/gitlab.spec.js]: again, could be nice to have a `testRemoteUser` in order to avoid duplication

**Comment by ercanucan** [comment]
@dlouzan looks nice! again, some small comments from my side

File: test/unit/authcache.spec.js (hunk 1 above)
Comment (dlouzan) [reply test/unit/authcache.spec.js]: I will do some refactoring in there

File: test/unit/gitlab.spec.js (hunk 3 above)
Comment (dlouzan) [reply test/unit/gitlab.spec.js]: Same as above

**Pull request merged**